File: QuickStarts/VB/EventBroker/Tests/Tests.EventBrokerExtension.csproj/Utility/MockBuilderContext.vb
'===============================================================================
' Microsoft patterns & practices
' Unity Application Block
'===============================================================================
' Copyright © Microsoft Corporation.  All rights reserved.
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
' OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
' LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
' FITNESS FOR A PARTICULAR PURPOSE.
'===============================================================================

Imports Microsoft.Practices.ObjectBuilder2

Namespace Tests.EventBrokerExtension.Utility
    Public Class MockBuilderContext
        Implements IBuilderContext
        Private _lifetime As ILifetimeContainer = New LifetimeContainer()
        Private _locator As IReadWriteLocator
        Private _originalBuildKey As Object = Nothing
        Private _persistentPolicies As IPolicyList = New PolicyList()
        Private _transientPolicies As IPolicyList
        Private _strategies As New StrategyChain()

        Private _buildKey As Object = Nothing
        Private _existing As Object = Nothing
        Private _buildComplete As Boolean

        Public Sub New()
            Me.New(New Locator())
        End Sub

        Public Sub New(ByVal locator As IReadWriteLocator)
            Me._locator = locator
        End Sub

        Public ReadOnly Property Lifetime() As ILifetimeContainer Implements IBuilderContext.Lifetime
            Get
                Return _lifetime
            End Get
        End Property

        Public ReadOnly Property Locator() As IReadWriteLocator Implements IBuilderContext.Locator
            Get
                Return _locator
            End Get
        End Property

        Public ReadOnly Property OriginalBuildKey() As Object Implements IBuilderContext.OriginalBuildKey
            Get
                Return _originalBuildKey
            End Get
        End Property

        Public ReadOnly Property PersistentPolicies() As IPolicyList Implements IBuilderContext.PersistentPolicies
            Get
                Return _persistentPolicies
            End Get
        End Property

        Public ReadOnly Property Policies() As IPolicyList Implements IBuilderContext.Policies
            Get
                Return _transientPolicies
            End Get
        End Property

        Public ReadOnly Property Strategies() As IStrategyChain Implements IBuilderContext.Strategies
            Get
                Return _strategies
            End Get
        End Property

        Public Property BuildKey() As Object Implements IBuilderContext.BuildKey
            Get
                Return _buildKey
            End Get
            Set(ByVal value As Object)
                _buildKey = value
            End Set
        End Property

        Public Property Existing() As Object Implements IBuilderContext.Existing
            Get
                Return _existing
            End Get
            Set(ByVal value As Object)
                _existing = value
            End Set
        End Property

        Public Property BuildComplete() As Boolean Implements IBuilderContext.BuildComplete
            Get
                Return _buildComplete
            End Get
            Set(ByVal value As Boolean)
                _buildComplete = value
            End Set
        End Property

        Public Function CloneForNewBuild(ByVal newBuildKey As Object, ByVal newExistingObject As Object) As IBuilderContext Implements IBuilderContext.CloneForNewBuild
            Dim newContext As New MockBuilderContext(Me._locator)
            newContext._strategies = _strategies
            newContext._persistentPolicies = _persistentPolicies
            newContext._transientPolicies = _transientPolicies
            newContext._lifetime = _lifetime
            newContext._originalBuildKey = newBuildKey
            newContext._buildKey = newBuildKey
            newContext._existing = newExistingObject
            Return newContext
        End Function

        Public Function ExecuteBuildUp(ByVal buildKey As Object, ByVal existing As Object)
            Me._buildKey = buildKey
            Me._existing = existing

            Return _strategies.ExecuteBuildUp(Me)
        End Function

        Public Function ExecuteTearDown(ByVal existing As Object) As Object
            Me._buildKey = Nothing
            Me._existing = existing

            _strategies.Reverse().ExecuteTearDown(Me)
            Return existing
        End Function
    End Class
End Namespace

File: Unity/Unity/Tests/TestSupport.Unity/MockBuilderContext.cs
//===============================================================================
// Microsoft patterns & practices
// Unity Application Block
//===============================================================================
// Copyright © Microsoft Corporation.  All rights reserved.
// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
// OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
// LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
// FITNESS FOR A PARTICULAR PURPOSE.
//===============================================================================

using System;
using System.Collections.Generic;
using Microsoft.Practices.ObjectBuilder2;

namespace Microsoft.Practices.Unity.TestSupport
{
    public class MockBuilderContext : IBuilderContext
    {
        ILifetimeContainer lifetime = new LifetimeContainer();
        NamedTypeBuildKey originalBuildKey = null;
        private IPolicyList persistentPolicies;
        IPolicyList policies;
        StrategyChain strategies = new StrategyChain();
        private CompositeResolverOverride resolverOverrides = new CompositeResolverOverride();

        private NamedTypeBuildKey buildKey = null;
        private object existing = null;
        private IRecoveryStack recoveryStack = new RecoveryStack();

        public MockBuilderContext()
        {
            this.persistentPolicies = new PolicyList();
            this.policies = new PolicyList(persistentPolicies);
        }

        public ILifetimeContainer Lifetime
        {
            get { return lifetime; }
        }

        public NamedTypeBuildKey OriginalBuildKey
        {
            get { return originalBuildKey; }
        }


        public IPolicyList PersistentPolicies
        {
            get { return persistentPolicies; }
        }

        public IPolicyList Policies
        {
            get { return policies; }
        }

        public IRecoveryStack RecoveryStack
        {
            get { return recoveryStack; }
        }

        public StrategyChain Strategies
        {
            get { return strategies; }
        }

        IStrategyChain IBuilderContext.Strategies
        {
            get { return strategies; }
        }


        public NamedTypeBuildKey BuildKey
        {
            get { return buildKey; }
            set { buildKey = value; }
        }

        public object Existing
        {
            get { return existing; }
            set { existing = value; }
        }

        public bool BuildComplete { get; set; }

        public object CurrentOperation { get; set; }

        public IBuilderContext ChildContext { get; set; }

        public void AddResolverOverrides(IEnumerable<ResolverOverride> newOverrides)
        {
            resolverOverrides.AddRange(newOverrides);
        }

        public IDependencyResolverPolicy GetOverriddenResolver(Type dependencyType)
        {
            return resolverOverrides.GetResolver(this, dependencyType);
        }

        public IBuilderContext CloneForNewBuild(NamedTypeBuildKey newBuildKey, object newExistingObject)
        {
            var newContext = new MockBuilderContext
                                 {
                                     strategies = strategies,
                                     persistentPolicies = persistentPolicies,
                                     policies = policies,
                                     lifetime = lifetime,
                                     originalBuildKey = buildKey,
                                     buildKey = newBuildKey,
                                     existing = newExistingObject
                                 };
            newContext.resolverOverrides.Add(resolverOverrides);

            return newContext;
        }

        /// <summary>
        /// A convenience method to do a new buildup operation on an existing context.
        /// </summary>
        /// <param name="newBuildKey">Key to use to build up.</param>
        /// <returns>Created object.</returns>
        public object NewBuildUp(NamedTypeBuildKey newBuildKey)
        {
            var clone = CloneForNewBuild(newBuildKey, null);
            return clone.Strategies.ExecuteBuildUp(clone);
        }

        /// <summary>
        /// A convenience method to do a new buildup operation on an existing context. This
        /// overload allows you to specify extra policies which will be in effect for the duration
        /// of the build.
        /// </summary>
        /// <param name="newBuildKey">Key defining what to build up.</param>
        /// <param name="childCustomizationBlock">A delegate that takes a <see cref="IBuilderContext"/>. This
        /// is invoked with the new child context before the build up process starts. This gives callers
        /// the opportunity to customize the context for the build process.</param>
        /// <returns>Created object.</returns>
        public object NewBuildUp(NamedTypeBuildKey newBuildKey, Action<IBuilderContext> childCustomizationBlock)
        {
            var newContext = new MockBuilderContext
            {
                strategies = strategies,
                persistentPolicies = persistentPolicies,
                policies = new PolicyList(persistentPolicies),
                lifetime = lifetime,
                originalBuildKey = buildKey,
                buildKey = newBuildKey,
                existing = null
            };
            newContext.resolverOverrides.Add(resolverOverrides);

            childCustomizationBlock(newContext);
            
            return strategies.ExecuteBuildUp(newContext);
        }

        public object ExecuteBuildUp(NamedTypeBuildKey buildKey, object existing)
        {
            this.BuildKey = buildKey;
            this.Existing = existing;

            return Strategies.ExecuteBuildUp(this);
        }

        public object ExecuteTearDown(object existing)
        {
            this.BuildKey = null;
            this.Existing = existing;

            Strategies.Reverse().ExecuteTearDown(this);
            return existing;
        }
    }
}
